{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary\n",
    "#  [Importing libraries](#Importing-libraries)\n",
    "#  [Initial configuration](#Initial-configuration)\n",
    "#  [Importing Excel spreadsheet](#Defining-functions)\n",
    "#  [Importing Excel spreadsheet](#Importing-Excel-spreadsheet)\n",
    "# [Setting up a local copy of the NCBI taxonomy database](#Setting-up-a-local-copy-of-the-NCBI-taxonomy-database)\n",
    "# [Geting NCBI TAX IDs](#Geting-NCBI-TAX-IDs)\n",
    "# [Creating dataframe with all microorganisms in the rows and the plastics in the columns](#Creating-dataframe-with-all-microorganisms-in-the-rows-and-the-plastics-in-the-columns)\n",
    "# [Filling the dataframe with plastic degrading data](#Filling-the-dataframe-with-plastic-degrading-data)\n",
    "# [Getting tree topology from NCBI and creating the tree](#Getting-tree-topology-from-NCBI-and-creating-the-tree)\n",
    "# [Creating Itol dataset files](#Creating-Itol-dataset-files)\n",
    "# [Downloading all available genomes for the plastic degraders](#Downloading-all-available-genomes-for-the-plastic-degraders)\n",
    "# [Creating itol dataset for microorganisms with available genomes](#Creating-itol-dataset-for-microorganisms-with-available-genomes)\n",
    "# [Blasting genes against genomes](#Blasting-genes-against-genomes)\n",
    "# [Creating itol datasets for genes](#Creating-itol-datasets-for-genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Importing-libraries'></a>\n",
    "# Importing libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import random, os, time\n",
    "from ete3 import NCBITaxa\n",
    "from ete3 import Tree\n",
    "import seaborn as sns\n",
    "import unicodedata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Initial-configuration'></a>\n",
    "# Initial configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#If running on a Slurm cluster, change account name and change run_on_slurm variable to True.\n",
    "account = \"uoa00354\"\n",
    "run_on_slurm = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Defining-functions'></a>\n",
    "# Defining functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_directory(directory_name):\n",
    "    if os.path.exists(directory_name) == False:\n",
    "        os.mkdir(directory_name)\n",
    "    \n",
    "def run_command(run_on_slurm, account, job_name, job_time, job_memory, job_cpus, job_command):\n",
    "    if run_on_slurm == False:\n",
    "        os.system(job_command)\n",
    "    if run_on_slurm == True:\n",
    "        f = open(\"run.sh\", \"w\")\n",
    "        f.write(\"\"\"#!/bin/bash\n",
    "\n",
    "#SBATCH --account=\"\"\"+str(account)+\"\"\"\n",
    "#SBATCH --job-name=\"\"\"+str(job_name)+\"\"\"\n",
    "#SBATCH --time=\"\"\"+str(job_time)+\"\"\"\n",
    "#SBATCH --mem=\"\"\"+str(job_memory)+\"\"\"\n",
    "#SBATCH --cpus-per-task=\"\"\"+str(job_cpus)+\"\"\"\n",
    "\n",
    "module load Python\n",
    "module load BLAST\n",
    "\n",
    "\"\"\"+job_command)\n",
    "        f.close()\n",
    "    \n",
    "        os.system(\"sbatch run.sh\")  \n",
    "        os.system(\"rm run.sh\")\n",
    "    else:\n",
    "        print(\"Please, set 'run_on_slurm' variable to True or False\")\n",
    "        \n",
    "def is_error(func, *args, **kw):\n",
    "    try:\n",
    "        func(*args, **kw)\n",
    "        return False\n",
    "    except Exception:\n",
    "        return True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Importing-Excel-spreadsheet'></a>\n",
    "# Importing Excel spreadsheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Microorganism</th>\n",
       "      <th>Tax ID</th>\n",
       "      <th>Plastic</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Enzyme</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Absidia sp.</td>\n",
       "      <td>0</td>\n",
       "      <td>PHB</td>\n",
       "      <td>Jeszeová, L., Puškárová, A., Bučková, M., Krak...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Absidia sp.</td>\n",
       "      <td>0</td>\n",
       "      <td>PLA</td>\n",
       "      <td>Jeszeová, L., Puškárová, A., Bučková, M., Krak...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Achromobacter xylosoxidans</td>\n",
       "      <td>0</td>\n",
       "      <td>HDPE</td>\n",
       "      <td>Kowalczyk, A., Chyc, M., Ryszka, P., &amp; Latowsk...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acidovorax avenae</td>\n",
       "      <td>0</td>\n",
       "      <td>PCL</td>\n",
       "      <td>Mergaert, J., Ruffieux, K., Bourban, C., Storm...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acidovorax avenae</td>\n",
       "      <td>0</td>\n",
       "      <td>PBSA</td>\n",
       "      <td>Mergaert, J., Ruffieux, K., Bourban, C., Storm...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Acidovorax delafieldii</td>\n",
       "      <td>0</td>\n",
       "      <td>PHBH</td>\n",
       "      <td>Morohoshi, T., Oi, T., Aiso, H., Suzuki, T., O...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Acidovorax delafieldii</td>\n",
       "      <td>0</td>\n",
       "      <td>PBSA</td>\n",
       "      <td>Uchida, H., Nakajima-Kambe, T., Shigeno-Akutsu...</td>\n",
       "      <td>Lipase</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Microorganism  Tax ID Plastic  \\\n",
       "0                 Absidia sp.       0     PHB   \n",
       "1                 Absidia sp.       0     PLA   \n",
       "2  Achromobacter xylosoxidans       0    HDPE   \n",
       "3           Acidovorax avenae       0     PCL   \n",
       "4           Acidovorax avenae       0    PBSA   \n",
       "5      Acidovorax delafieldii       0    PHBH   \n",
       "6      Acidovorax delafieldii       0    PBSA   \n",
       "\n",
       "                                                 Ref  Enzyme Gene Sequence  \n",
       "0  Jeszeová, L., Puškárová, A., Bučková, M., Krak...      No   No      NaN  \n",
       "1  Jeszeová, L., Puškárová, A., Bučková, M., Krak...      No   No      NaN  \n",
       "2  Kowalczyk, A., Chyc, M., Ryszka, P., & Latowsk...      No   No      NaN  \n",
       "3  Mergaert, J., Ruffieux, K., Bourban, C., Storm...      No   No      NaN  \n",
       "4  Mergaert, J., Ruffieux, K., Bourban, C., Storm...      No   No      NaN  \n",
       "5  Morohoshi, T., Oi, T., Aiso, H., Suzuki, T., O...      No   No      NaN  \n",
       "6  Uchida, H., Nakajima-Kambe, T., Shigeno-Akutsu...  Lipase   No      NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_excel(\"Degraders_list.xlsx\")\n",
    "#Visualize part of the dataset.\n",
    "df.head(7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Setting-up-a-local-copy-of-the-NCBI-taxonomy-database'></a>\n",
    "# Setting up a local copy of the NCBI taxonomy database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi = NCBITaxa()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Geting-NCBI-TAX-IDs'></a>\n",
    "# Geting NCBI TAX IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/vmar559/PyPackages/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \"\"\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Microorganism</th>\n",
       "      <th>Tax ID</th>\n",
       "      <th>Plastic</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Enzyme</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Absidia sp.</td>\n",
       "      <td>1982014</td>\n",
       "      <td>PHB</td>\n",
       "      <td>Jeszeová, L., Puškárová, A., Bučková, M., Krak...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Absidia sp.</td>\n",
       "      <td>1982014</td>\n",
       "      <td>PLA</td>\n",
       "      <td>Jeszeová, L., Puškárová, A., Bučková, M., Krak...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Achromobacter xylosoxidans</td>\n",
       "      <td>85698</td>\n",
       "      <td>HDPE</td>\n",
       "      <td>Kowalczyk, A., Chyc, M., Ryszka, P., &amp; Latowsk...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acidovorax avenae</td>\n",
       "      <td>80867</td>\n",
       "      <td>PCL</td>\n",
       "      <td>Mergaert, J., Ruffieux, K., Bourban, C., Storm...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acidovorax avenae</td>\n",
       "      <td>80867</td>\n",
       "      <td>PBSA</td>\n",
       "      <td>Mergaert, J., Ruffieux, K., Bourban, C., Storm...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Acidovorax delafieldii</td>\n",
       "      <td>47920</td>\n",
       "      <td>PHBH</td>\n",
       "      <td>Morohoshi, T., Oi, T., Aiso, H., Suzuki, T., O...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Acidovorax delafieldii</td>\n",
       "      <td>47920</td>\n",
       "      <td>PBSA</td>\n",
       "      <td>Uchida, H., Nakajima-Kambe, T., Shigeno-Akutsu...</td>\n",
       "      <td>Lipase</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Microorganism   Tax ID Plastic  \\\n",
       "0                 Absidia sp.  1982014     PHB   \n",
       "1                 Absidia sp.  1982014     PLA   \n",
       "2  Achromobacter xylosoxidans    85698    HDPE   \n",
       "3           Acidovorax avenae    80867     PCL   \n",
       "4           Acidovorax avenae    80867    PBSA   \n",
       "5      Acidovorax delafieldii    47920    PHBH   \n",
       "6      Acidovorax delafieldii    47920    PBSA   \n",
       "\n",
       "                                                 Ref  Enzyme Gene Sequence  \n",
       "0  Jeszeová, L., Puškárová, A., Bučková, M., Krak...      No   No      NaN  \n",
       "1  Jeszeová, L., Puškárová, A., Bučková, M., Krak...      No   No      NaN  \n",
       "2  Kowalczyk, A., Chyc, M., Ryszka, P., & Latowsk...      No   No      NaN  \n",
       "3  Mergaert, J., Ruffieux, K., Bourban, C., Storm...      No   No      NaN  \n",
       "4  Mergaert, J., Ruffieux, K., Bourban, C., Storm...      No   No      NaN  \n",
       "5  Morohoshi, T., Oi, T., Aiso, H., Suzuki, T., O...      No   No      NaN  \n",
       "6  Uchida, H., Nakajima-Kambe, T., Shigeno-Akutsu...  Lipase   No      NaN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#If it returns an error, just wait, it will get the dataframe anyway.\n",
    "for index, row in df.iterrows():\n",
    "    if row[\"Tax ID\"] == 0:\n",
    "        name2taxid = ncbi.get_name_translator([row[\"Microorganism\"]])\n",
    "        df[\"Tax ID\"][row.name] = str(name2taxid.values()).replace(\"[\",\"\").replace(\"]\",\"\").split(\"(\")[1].split(\")\")[0]\n",
    "df.head(7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-dataframe-with-all-microorganisms-in-the-rows-and-the-plastics-in-the-columns'></a>\n",
    "# Creating dataframe with all microorganisms in the rows and the plastics in the columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BTA copolyester</th>\n",
       "      <th>Ecoflex</th>\n",
       "      <th>HDPE</th>\n",
       "      <th>LDPE</th>\n",
       "      <th>LDPE Blend</th>\n",
       "      <th>Mater-Bi</th>\n",
       "      <th>Nylon</th>\n",
       "      <th>Nylon 12</th>\n",
       "      <th>Nylon 4</th>\n",
       "      <th>Nylon 6</th>\n",
       "      <th>...</th>\n",
       "      <th>PTMS</th>\n",
       "      <th>PTS</th>\n",
       "      <th>PU</th>\n",
       "      <th>PU Blend</th>\n",
       "      <th>PVA-LLDPE</th>\n",
       "      <th>PVC</th>\n",
       "      <th>PVC Blend</th>\n",
       "      <th>Poly(3HB-co-4HB)</th>\n",
       "      <th>Polyether sulfone</th>\n",
       "      <th>Sky-Green</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     BTA copolyester  Ecoflex  HDPE  LDPE  LDPE Blend  Mater-Bi  Nylon  \\\n",
       "232                0        0     0     0           0         0      0   \n",
       "239                0        0     0     0           0         0      0   \n",
       "285                0        0     0     0           0         0      0   \n",
       "287                0        0     0     0           0         0      0   \n",
       "292                0        0     0     0           0         0      0   \n",
       "294                0        0     0     0           0         0      0   \n",
       "300                0        0     0     0           0         0      0   \n",
       "\n",
       "     Nylon 12  Nylon 4  Nylon 6  ...  PTMS  PTS  PU  PU Blend  PVA-LLDPE  PVC  \\\n",
       "232         0        0        0  ...     0    0   0         0          0    0   \n",
       "239         0        0        0  ...     0    0   0         0          0    0   \n",
       "285         0        0        0  ...     0    0   0         0          0    0   \n",
       "287         0        0        0  ...     0    0   0         0          0    0   \n",
       "292         0        0        0  ...     0    0   0         0          0    0   \n",
       "294         0        0        0  ...     0    0   0         0          0    0   \n",
       "300         0        0        0  ...     0    0   0         0          0    0   \n",
       "\n",
       "     PVC Blend  Poly(3HB-co-4HB)  Polyether sulfone  Sky-Green  \n",
       "232          0                 0                  0          0  \n",
       "239          0                 0                  0          0  \n",
       "285          0                 0                  0          0  \n",
       "287          0                 0                  0          0  \n",
       "292          0                 0                  0          0  \n",
       "294          0                 0                  0          0  \n",
       "300          0                 0                  0          0  \n",
       "\n",
       "[7 rows x 63 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plastic_types = df['Plastic'].unique()\n",
    "tax_ids = df['Tax ID'].unique()\n",
    "df2 = pd.DataFrame(0, index=tax_ids, columns=plastic_types)\n",
    "df2.sort_index(inplace=True)\n",
    "df2.sort_index(axis=1, inplace=True)\n",
    "df2.head(7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Filling-the-dataframe-with-plastic-degrading-data'></a>\n",
    "# Filling the dataframe with plastic degrading data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BTA copolyester</th>\n",
       "      <th>Ecoflex</th>\n",
       "      <th>HDPE</th>\n",
       "      <th>LDPE</th>\n",
       "      <th>LDPE Blend</th>\n",
       "      <th>Mater-Bi</th>\n",
       "      <th>Nylon</th>\n",
       "      <th>Nylon 12</th>\n",
       "      <th>Nylon 4</th>\n",
       "      <th>Nylon 6</th>\n",
       "      <th>...</th>\n",
       "      <th>PTMS</th>\n",
       "      <th>PTS</th>\n",
       "      <th>PU</th>\n",
       "      <th>PU Blend</th>\n",
       "      <th>PVA-LLDPE</th>\n",
       "      <th>PVC</th>\n",
       "      <th>PVC Blend</th>\n",
       "      <th>Poly(3HB-co-4HB)</th>\n",
       "      <th>Polyether sulfone</th>\n",
       "      <th>Sky-Green</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     BTA copolyester  Ecoflex  HDPE  LDPE  LDPE Blend  Mater-Bi  Nylon  \\\n",
       "232                0        0     0     0           0         0      0   \n",
       "239                0        0     0     0           0         0      0   \n",
       "285                0        0     0     0           0         0      0   \n",
       "287                0        0     1     1           0         0      0   \n",
       "292                0        0     0     0           0         0      0   \n",
       "294                0        0     1     0           0         0      0   \n",
       "300                0        0     0     0           0         0      0   \n",
       "\n",
       "     Nylon 12  Nylon 4  Nylon 6  ...  PTMS  PTS  PU  PU Blend  PVA-LLDPE  PVC  \\\n",
       "232         0        0        0  ...     0    0   0         0          0    0   \n",
       "239         0        0        0  ...     0    0   0         0          0    0   \n",
       "285         0        0        0  ...     0    0   0         0          0    0   \n",
       "287         0        0        0  ...     0    0   1         1          0    0   \n",
       "292         0        0        0  ...     0    0   0         0          0    0   \n",
       "294         0        0        0  ...     0    0   0         0          0    1   \n",
       "300         0        0        0  ...     0    0   0         0          0    0   \n",
       "\n",
       "     PVC Blend  Poly(3HB-co-4HB)  Polyether sulfone  Sky-Green  \n",
       "232          0                 0                  0          0  \n",
       "239          0                 0                  0          0  \n",
       "285          0                 0                  0          0  \n",
       "287          0                 0                  0          0  \n",
       "292          0                 0                  0          0  \n",
       "294          0                 0                  0          0  \n",
       "300          0                 0                  0          0  \n",
       "\n",
       "[7 rows x 63 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references_dict = {}\n",
    "for index, row in df.iterrows():\n",
    "    plastic = row['Plastic']\n",
    "    tax_id = row['Tax ID']\n",
    "    reference = str(row['Ref'].strip())\n",
    "    normalized = unicodedata.normalize('NFD', reference)\n",
    "    reference = u\"\".join([c for c in normalized if not unicodedata.combining(c)])+\"<br><br>\"\n",
    "    df2.at[tax_id, plastic] = 1\n",
    "    if tax_id not in references_dict.keys():\n",
    "        references_dict[tax_id] = reference\n",
    "    else:\n",
    "        if reference not in references_dict[tax_id]:\n",
    "            references_dict[tax_id] += reference\n",
    "df2.head(7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Getting-tree-topology-from-NCBI-and-creating-the-tree'></a>\n",
    "# Getting tree topology from NCBI and creating the tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = df2.index.values\n",
    "t = ncbi.get_topology(ids, intermediate_nodes=True)\n",
    "t.write(format=8, outfile=\"tree.nw\")\n",
    "ids = df2.index.values\n",
    "t = ncbi.get_topology(ids, intermediate_nodes=True)\n",
    "t.write(format=8, outfile=\"tree.nw\")\n",
    "#http://phylogeny.lirmm.fr/phylo_cgi/data_converter.cgi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-Itol-dataset-files'></a>\n",
    "# Creating Itol dataset files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = lambda: random.randint(0,255) #Creates random hexadecimal color codes\n",
    "create_directory(\"itol_datasets\")\n",
    "\n",
    "for column in df2.columns:\n",
    "    f = open(\"itol_datasets/\"+column.rstrip()+\".txt\", \"w\")\n",
    "    f.write(\"\"\"DATASET_BINARY\n",
    "SEPARATOR COMMA\n",
    "DATASET_LABEL,\"\"\"+column+\"\"\"\n",
    "COLOR,\"\"\"+'#%02X%02X%02X' % (r(),r(),r())+\"\"\"\n",
    "FIELD_SHAPES,1\n",
    "FIELD_LABELS,f1\n",
    "DATA\"\"\")\n",
    "    for index, row in df2.iterrows():\n",
    "        tax_id = str(index)\n",
    "        feature_binary = str(row[column])\n",
    "        f.write(\"\\n\"+tax_id+\",\"+feature_binary)\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Downloading-all-available-genomes-for-the-plastic-degraders'></a>\n",
    "# Downloading all available genomes for the plastic degraders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "for tax_id in tax_ids:\n",
    "    job_name = str(tax_id)+\"_download\"\n",
    "    job_time = \"05:00:00\"\n",
    "    job_memory = 100 #Amount of ram memory in MB\n",
    "    job_cpus = 1\n",
    "    job_command = \"ncbi-genome-download -s genbank --format fasta --taxid \"+str(tax_id)+\" -r 5 -o genomes/\"+str(tax_id)+\"/ fungi,bacteria\"\n",
    "    run_command(run_on_slurm, account, job_name, job_time, job_memory, job_cpus, job_command)\n",
    "    time.sleep(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-itol-dataset-for-microorganisms-with-available-genomes'></a>\n",
    "# Creating itol dataset for microorganisms with available genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "create_directory(\"itol_datasets_genomes\")\n",
    "directories_list = os.listdir(\"genomes\")\n",
    "\n",
    "f = open(\"itol_datasets_genomes/Available_genomes.txt\", \"w\")\n",
    "f.write(\"\"\"DATASET_BINARY\n",
    "SEPARATOR COMMA\n",
    "DATASET_LABEL,Available genomes\n",
    "COLOR,#FF0000\n",
    "FIELD_SHAPES,1\n",
    "FIELD_LABELS,f1\n",
    "DATA\"\"\")\n",
    "\n",
    "for index, row in df2.iterrows():\n",
    "    tax_id = str(index)\n",
    "    if str(index) in directories_list:\n",
    "        f.write(\"\\n\"+tax_id+\",\"+\"1\")\n",
    "    else:\n",
    "        f.write(\"\\n\"+tax_id+\",\"+\"0\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Blasting-genes-against-genomes'></a>\n",
    "# Blasting genes against genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "\n",
    "#Adding the TAX ID to all sequence names\n",
    "create_directory(\"genomes_database\")\n",
    "os.system(\"touch genomes_database/genomes.fasta\")\n",
    "for directory in directories_list:\n",
    "    os.system(\"zcat genomes/\"+directory+\"/genbank/*/*/*genomic.fna.gz | sed s/'^>'/'>\"+directory+\"_\"+\"'/g >> genomes_database/genomes.fasta\")\n",
    "\n",
    "#Creating BLAST database\n",
    "os.system(\"makeblastdb -in genomes_database/genomes.fasta -dbtype nucl\")\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "#Sending BLAST jobs\n",
    "create_directory(\"blasts\")\n",
    "job_time = \"01:00:00\"\n",
    "job_memory = 512 #Amount of ram memory in MB\n",
    "job_cpus = 1\n",
    "for index, row in df.iterrows():\n",
    "    sequences = row[\"Sequence\"]\n",
    "    if pd.notna(sequences) == True:\n",
    "        for n in range(0, len(sequences.split(\";\"))):\n",
    "            species_name = tax_id_to_organism_name_dict[row[\"Tax ID\"]].replace(\" \",\"_\")\n",
    "            job_name = str(row[\"Tax ID\"])+\"_\"+str(n)\n",
    "            job_command = \"echo '\"+sequences.split(\";\")[n]+\"' | tblastn -query - -db genomes_database/genomes.fasta -out blasts/\"+species_name+\"_\"+row[\"Enzyme\"].replace(\" \",\"_\")+\"_\"+str(n)+\".tsv -outfmt 6\"\n",
    "            run_command(run_on_slurm, account, job_name, job_time, job_memory, job_cpus, job_command)\n",
    "            time.sleep(2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-itol-datasets-for-all-genes'></a>\n",
    "# Creating itol datasets for all genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = lambda: random.randint(0,255) #Creates random hexadecimal color codes\n",
    "\n",
    "blast_outputs_list = os.listdir(\"blasts\")\n",
    "for file in blast_outputs_list:\n",
    "    file_blast = open(\"blasts/\"+file)\n",
    "    blast = file_blast.readlines()\n",
    "    file_itol = open(\"itol_datasets_genomes/\"+file.replace(\"tsv\", \"txt\"), \"w\")\n",
    "    counting_dict = {}\n",
    "    for line in blast:\n",
    "        qseqid = line.rstrip().split(\"\\t\")[0]\n",
    "        sseqid = line.rstrip().split(\"\\t\")[1].split(\"_\")[0]\n",
    "        pident = float(line.rstrip().split(\"\\t\")[2])\n",
    "        length = int(line.rstrip().split(\"\\t\")[3])\n",
    "        mismatch = int(line.rstrip().split(\"\\t\")[4])\n",
    "        gapopen = int(line.rstrip().split(\"\\t\")[5])\n",
    "        qstart = int(line.rstrip().split(\"\\t\")[6])\n",
    "        qend = float(line.rstrip().split(\"\\t\")[7])\n",
    "        sstart = float(line.rstrip().split(\"\\t\")[8])\n",
    "        send = float(line.rstrip().split(\"\\t\")[9])\n",
    "        evalue = float(line.rstrip().split(\"\\t\")[10])\n",
    "        bitscore = float(line.rstrip().split(\"\\t\")[11])\n",
    "        if evalue <= 1e-06 and pident >= 25:\n",
    "            if sseqid not in counting_dict:\n",
    "                counting_dict[sseqid] = 1\n",
    "            else:\n",
    "                counting_dict[sseqid] += 1\n",
    "                \n",
    "    file_itol.write(\"\"\"DATASET_MULTIBAR\n",
    "SEPARATOR TAB\n",
    "DATASET_LABEL   \"\"\"+file.replace('.tsv', '').replace('_', ' ')+\"\"\"\n",
    "COLOR   \"\"\"+'#%02X%02X%02X' % (r(),r(),r())+\"\"\"\n",
    "FIELD_COLORS    \"\"\"+'#%02X%02X%02X' % (r(),r(),r())+\"\"\"\n",
    "FIELD_LABELS    \"\"\"+file.replace('.tsv', '').replace('_', ' ')+\"\"\"\n",
    "DATA\"\"\")    \n",
    "    for index, row in df2.iterrows():\n",
    "        if str(index) in counting_dict:\n",
    "            file_itol.write(\"\\n\"+str(index)+\"\\t\"+str(counting_dict[str(index)]))\n",
    "        else:\n",
    "            file_itol.write(\"\\n\"+str(index)+\"\\t0\")\n",
    "    file_blast.close()\n",
    "    file_itol.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-one-light-color-for-each-phylum'></a>\n",
    "# Creating one light color for each phylum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylum_color_dict = {}\n",
    "phylum_color_list = []\n",
    "for node in t.traverse(\"postorder\"):\n",
    "    if \"'phylum'\" in str(ncbi.get_rank([int(node.name)])):\n",
    "        phylum_name = ncbi.get_taxid_translator([int(node.name)])[int(node.name)]\n",
    "        phylum_color_list.append(phylum_name)\n",
    "phylum_colors = sns.color_palette(\"hls\", (len(phylum_color_list))).as_hex()\n",
    "n = 0\n",
    "for item in phylum_color_list:\n",
    "    phylum_color_dict[item] = phylum_colors[n]\n",
    "    n += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-itol-datasets-for-phylum-labels'></a>\n",
    "# Creating itol datasets for phylum labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(\"itol_datasets/Phylum_labels.txt\", \"w\")\n",
    "f.write(\"\"\"DATASET_TEXT\n",
    "SEPARATOR COMMA\n",
    "DATASET_LABEL,Phylum Labels\n",
    "COLOR,#ff0000\n",
    "MARGIN,0\n",
    "SHOW_INTERNAL,1\n",
    "LABEL_ROTATION,0\n",
    "STRAIGHT_LABELS,0\n",
    "ALIGN_TO_TREE,0\n",
    "SIZE_FACTOR,0\n",
    "DATA\n",
    "\"\"\")\n",
    "for node in t.traverse(\"postorder\"):\n",
    "    if \"'phylum'\" in str(ncbi.get_rank([int(node.name)])):\n",
    "        phylum_name = ncbi.get_taxid_translator([int(node.name)])[int(node.name)]\n",
    "        f.write(\"\\ninode\"+node.name+\",\"+phylum_name+\",0,\"+phylum_color_dict[phylum_name]+\",bold,10,0\")      \n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-itol-datasets-for-phylum-colored-ranges'></a>\n",
    "# Creating itol datasets for phylum colored ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(\"itol_datasets/Phylum_colors.txt\", \"w\")\n",
    "f.write(\"\"\"TREE_COLORS\n",
    "SEPARATOR COMMA\n",
    "DATA\n",
    "\"\"\")\n",
    "for node in t.traverse(\"postorder\"):\n",
    "    if \"'phylum'\" in str(ncbi.get_rank([int(node.name)])):\n",
    "        phylum_name = ncbi.get_taxid_translator([int(node.name)])[int(node.name)]\n",
    "        f.write(\"\\ninode\"+node.name+\",range,\"+phylum_color_dict[phylum_name]+\",\"+phylum_name)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-itol-datasets-for-references'></a>\n",
    "# Creating itol datasets for references"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(\"itol_datasets/Popup_info.txt\", \"w\")\n",
    "f.write(\"\"\"POPUP_INFO\n",
    "SEPARATOR TAB\n",
    "DATA\n",
    "\"\"\")\n",
    "for index, row in df2.iterrows():\n",
    "    organism_name = ncbi.get_taxid_translator([index])[index]\n",
    "    html_string = \"<h1>\"+organism_name+\"</h1><br>NCBI ID = \"+str(index)+\"<br><br>References:<br><br>\"+references_dict[index]\n",
    "    f.write(\"\\n\"+str(index)+\"\\t\"+organism_name+\"\\t\"+html_string)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-tree-with-all-species'></a>\n",
    "# Creating tree with all species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "all_species_file = open(\"all_species_tax_ids.txt\")\n",
    "all_species = all_species_file.readlines()\n",
    "all_bacteria_and_fungi = []\n",
    "for item in all_species:\n",
    "    tax_id = item.rstrip()\n",
    "    if is_error(ncbi.get_lineage, tax_id) == False:\n",
    "        lineage = ncbi.get_lineage(tax_id)\n",
    "        fungi = 0\n",
    "        eukaryota = 0\n",
    "        for sub_item in lineage:\n",
    "            if \"Eukaryota\" in ncbi.get_taxid_translator([str(sub_item)])[sub_item]:\n",
    "                eukaryota = 1\n",
    "            if \"Fungi\" in ncbi.get_taxid_translator([str(sub_item)])[sub_item]:\n",
    "                fungi = 1\n",
    "        if eukaryota == 0:\n",
    "            all_bacteria_and_fungi.append(tax_id)\n",
    "        else:\n",
    "            if fungi == 1:\n",
    "                all_bacteria_and_fungi.append(tax_id)\n",
    "'''\n",
    "all_bacteria_and_fungi_subsample = random.sample(all_bacteria_and_fungi, 3000)                              \n",
    "\n",
    "for item in ids:\n",
    "    all_bacteria_and_fungi_subsample.append(str(item))\n",
    "t_all_species = ncbi.get_topology(all_bacteria_and_fungi_subsample, intermediate_nodes=True)\n",
    "t_all_species.write(format=8, outfile=\"tree_all_species.nw\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-one-light-color-for-each-phylum-for-all-species-tree'></a>\n",
    "# Creating one light color for each phylum for all species tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylum_color_dict_all_species = {}\n",
    "phylum_color_list_all_species = []\n",
    "for node in t_all_species.traverse(\"postorder\"):\n",
    "    if \"'phylum'\" in str(ncbi.get_rank([int(node.name)])):\n",
    "        phylum_name = ncbi.get_taxid_translator([int(node.name)])[int(node.name)]\n",
    "        phylum_color_list_all_species.append(phylum_name)\n",
    "phylum_colors = sns.color_palette(\"hls\", (len(phylum_color_list_all_species))).as_hex()\n",
    "n = 0\n",
    "for item in phylum_color_list_all_species:\n",
    "    phylum_color_dict_all_species[item] = phylum_colors[n]\n",
    "    n += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-itol-datasets-for-phylum-colored-ranges-for-all-species-tree'></a>\n",
    "# Creating itol datasets for phylum colored ranges for all species tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(\"itol_datasets/Phylum_colors_all_species.txt\", \"w\")\n",
    "f.write(\"\"\"TREE_COLORS\n",
    "SEPARATOR COMMA\n",
    "DATA\n",
    "\"\"\")\n",
    "for node in t_all_species.traverse(\"postorder\"):\n",
    "    if \"'phylum'\" in str(ncbi.get_rank([int(node.name)])):\n",
    "        phylum_name = ncbi.get_taxid_translator([int(node.name)])[int(node.name)]\n",
    "        f.write(\"\\ninode\"+node.name+\",range,\"+phylum_color_dict_all_species[phylum_name]+\",\"+phylum_name)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-itol-datasets-for-phylum-labels-for-all-species-tree'></a>\n",
    "# Creating itol datasets for phylum labels for all species tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(\"itol_datasets/Phylum_labels_all_species.txt\", \"w\")\n",
    "f.write(\"\"\"DATASET_TEXT\n",
    "SEPARATOR COMMA\n",
    "DATASET_LABEL,Phylum Labels\n",
    "COLOR,#ff0000\n",
    "MARGIN,0\n",
    "SHOW_INTERNAL,1\n",
    "LABEL_ROTATION,0\n",
    "STRAIGHT_LABELS,0\n",
    "ALIGN_TO_TREE,0\n",
    "SIZE_FACTOR,0\n",
    "DATA\n",
    "\"\"\")\n",
    "for node in t_all_species.traverse(\"postorder\"):\n",
    "    if \"'phylum'\" in str(ncbi.get_rank([int(node.name)])):\n",
    "        phylum_name = ncbi.get_taxid_translator([int(node.name)])[int(node.name)]\n",
    "        f.write(\"\\ninode\"+node.name+\",\"+phylum_name+\",0,\"+phylum_color_dict_all_species[phylum_name]+\",bold,10,0\")      \n",
    "f.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
