{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary\n",
    "#  [Importing libraries](#Importing-libraries)\n",
    "#  [Initial configuration](#Initial-configuration)\n",
    "#  [Importing Excel spreadsheet](#Defining-functions)\n",
    "#  [Importing Excel spreadsheet](#Importing-Excel-spreadsheet)\n",
    "# [Setting up a local copy of the NCBI taxonomy database](#Setting-up-a-local-copy-of-the-NCBI-taxonomy-database)\n",
    "# [Geting NCBI TAX IDs](#Geting-NCBI-TAX-IDs)\n",
    "# [Creating dataframe with all microorganisms in the rows and the plastics in the columns](#Creating-dataframe-with-all-microorganisms-in-the-rows-and-the-plastics-in-the-columns)\n",
    "# [Filling the dataframe with plastic degrading data](#Filling-the-dataframe-with-plastic-degrading-data)\n",
    "# [Getting tree topology from NCBI and creating the tree](#Getting-tree-topology-from-NCBI-and-creating-the-tree)\n",
    "# [Creating Itol dataset files](#Creating-Itol-dataset-files)\n",
    "# [Downloading all available genomes for the plastic degraders](#Downloading-all-available-genomes-for-the-plastic-degraders)\n",
    "# [Creating itol dataset for microorganisms with available genomes](#Creating-itol-dataset-for-microorganisms-with-available-genomes)\n",
    "# [Blasting genes against genomes](#Blasting-genes-against-genomes)\n",
    "# [Creating itol datasets for genes](#Creating-itol-datasets-for-genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Importing-libraries'></a>\n",
    "# Importing libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import random, os, time\n",
    "from ete3 import NCBITaxa\n",
    "from ete3 import Tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Initial-configuration'></a>\n",
    "# Initial configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#If running on a Slurm cluster, change account name and change run_on_slurm variable to True.\n",
    "account = \"uoa00354\"\n",
    "run_on_slurm = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Defining-functions'></a>\n",
    "# Defining functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_directory(directory_name):\n",
    "    if os.path.exists(directory_name) == False:\n",
    "        os.mkdir(directory_name)\n",
    "    \n",
    "def run_command(run_on_slurm, account, job_name, job_time, job_memory, job_cpus, job_command):\n",
    "    if run_on_slurm == False:\n",
    "        os.system(job_command)\n",
    "    if run_on_slurm == True:\n",
    "        f = open(\"run.sh\", \"w\")\n",
    "        f.write(\"\"\"#!/bin/bash\n",
    "\n",
    "#SBATCH --account=\"\"\"+str(account)+\"\"\"\n",
    "#SBATCH --job-name=\"\"\"+str(job_name)+\"\"\"\n",
    "#SBATCH --time=\"\"\"+str(job_time)+\"\"\"\n",
    "#SBATCH --mem=\"\"\"+str(job_memory)+\"\"\"\n",
    "#SBATCH --cpus-per-task=\"\"\"+str(job_cpus)+\"\"\"\n",
    "\n",
    "module load Python\n",
    "module load BLAST\n",
    "\n",
    "\"\"\"+job_command)\n",
    "        f.close()\n",
    "    \n",
    "        os.system(\"sbatch run.sh\")  \n",
    "        os.system(\"rm run.sh\")\n",
    "    else:\n",
    "        print(\"Please, set 'run_on_slurm' variable to True or False\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Importing-Excel-spreadsheet'></a>\n",
    "# Importing Excel spreadsheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Microorganism</th>\n",
       "      <th>Tax ID</th>\n",
       "      <th>Plastic</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Enzyme</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Absidia sp.</td>\n",
       "      <td>0</td>\n",
       "      <td>PHB</td>\n",
       "      <td>Jeszeová, L., Puškárová, A., Bučková, M., Krak...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Absidia sp.</td>\n",
       "      <td>0</td>\n",
       "      <td>PLA</td>\n",
       "      <td>Jeszeová, L., Puškárová, A., Bučková, M., Krak...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Achromobacter xylosoxidans</td>\n",
       "      <td>0</td>\n",
       "      <td>HDPE</td>\n",
       "      <td>Kowalczyk, A., Chyc, M., Ryszka, P., &amp; Latowsk...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acidovorax avenae</td>\n",
       "      <td>0</td>\n",
       "      <td>PCL</td>\n",
       "      <td>Mergaert, J., Ruffieux, K., Bourban, C., Storm...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acidovorax avenae</td>\n",
       "      <td>0</td>\n",
       "      <td>PBSA</td>\n",
       "      <td>Mergaert, J., Ruffieux, K., Bourban, C., Storm...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Acidovorax delafieldii</td>\n",
       "      <td>0</td>\n",
       "      <td>PHBH</td>\n",
       "      <td>Morohoshi, T., Oi, T., Aiso, H., Suzuki, T., O...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Acidovorax delafieldii</td>\n",
       "      <td>0</td>\n",
       "      <td>PBSA</td>\n",
       "      <td>Uchida, H., Nakajima-Kambe, T., Shigeno-Akutsu...</td>\n",
       "      <td>Lipase</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Microorganism  Tax ID Plastic  \\\n",
       "0                 Absidia sp.       0     PHB   \n",
       "1                 Absidia sp.       0     PLA   \n",
       "2  Achromobacter xylosoxidans       0    HDPE   \n",
       "3           Acidovorax avenae       0     PCL   \n",
       "4           Acidovorax avenae       0    PBSA   \n",
       "5      Acidovorax delafieldii       0    PHBH   \n",
       "6      Acidovorax delafieldii       0    PBSA   \n",
       "\n",
       "                                                 Ref  Enzyme Gene Sequence  \n",
       "0  Jeszeová, L., Puškárová, A., Bučková, M., Krak...      No   No      NaN  \n",
       "1  Jeszeová, L., Puškárová, A., Bučková, M., Krak...      No   No      NaN  \n",
       "2  Kowalczyk, A., Chyc, M., Ryszka, P., & Latowsk...      No   No      NaN  \n",
       "3  Mergaert, J., Ruffieux, K., Bourban, C., Storm...      No   No      NaN  \n",
       "4  Mergaert, J., Ruffieux, K., Bourban, C., Storm...      No   No      NaN  \n",
       "5  Morohoshi, T., Oi, T., Aiso, H., Suzuki, T., O...      No   No      NaN  \n",
       "6  Uchida, H., Nakajima-Kambe, T., Shigeno-Akutsu...  Lipase   No      NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_excel(\"Degraders_list.xlsx\")\n",
    "#Visualize part of the dataset.\n",
    "df.head(7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Setting-up-a-local-copy-of-the-NCBI-taxonomy-database'></a>\n",
    "# Setting up a local copy of the NCBI taxonomy database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi = NCBITaxa()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Geting-NCBI-TAX-IDs'></a>\n",
    "# Geting NCBI TAX IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/vmar559/PyPackages/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \"\"\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Microorganism</th>\n",
       "      <th>Tax ID</th>\n",
       "      <th>Plastic</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Enzyme</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Absidia sp.</td>\n",
       "      <td>1982014</td>\n",
       "      <td>PHB</td>\n",
       "      <td>Jeszeová, L., Puškárová, A., Bučková, M., Krak...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Absidia sp.</td>\n",
       "      <td>1982014</td>\n",
       "      <td>PLA</td>\n",
       "      <td>Jeszeová, L., Puškárová, A., Bučková, M., Krak...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Achromobacter xylosoxidans</td>\n",
       "      <td>85698</td>\n",
       "      <td>HDPE</td>\n",
       "      <td>Kowalczyk, A., Chyc, M., Ryszka, P., &amp; Latowsk...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acidovorax avenae</td>\n",
       "      <td>80867</td>\n",
       "      <td>PCL</td>\n",
       "      <td>Mergaert, J., Ruffieux, K., Bourban, C., Storm...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acidovorax avenae</td>\n",
       "      <td>80867</td>\n",
       "      <td>PBSA</td>\n",
       "      <td>Mergaert, J., Ruffieux, K., Bourban, C., Storm...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Acidovorax delafieldii</td>\n",
       "      <td>47920</td>\n",
       "      <td>PHBH</td>\n",
       "      <td>Morohoshi, T., Oi, T., Aiso, H., Suzuki, T., O...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Acidovorax delafieldii</td>\n",
       "      <td>47920</td>\n",
       "      <td>PBSA</td>\n",
       "      <td>Uchida, H., Nakajima-Kambe, T., Shigeno-Akutsu...</td>\n",
       "      <td>Lipase</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Microorganism   Tax ID Plastic  \\\n",
       "0                 Absidia sp.  1982014     PHB   \n",
       "1                 Absidia sp.  1982014     PLA   \n",
       "2  Achromobacter xylosoxidans    85698    HDPE   \n",
       "3           Acidovorax avenae    80867     PCL   \n",
       "4           Acidovorax avenae    80867    PBSA   \n",
       "5      Acidovorax delafieldii    47920    PHBH   \n",
       "6      Acidovorax delafieldii    47920    PBSA   \n",
       "\n",
       "                                                 Ref  Enzyme Gene Sequence  \n",
       "0  Jeszeová, L., Puškárová, A., Bučková, M., Krak...      No   No      NaN  \n",
       "1  Jeszeová, L., Puškárová, A., Bučková, M., Krak...      No   No      NaN  \n",
       "2  Kowalczyk, A., Chyc, M., Ryszka, P., & Latowsk...      No   No      NaN  \n",
       "3  Mergaert, J., Ruffieux, K., Bourban, C., Storm...      No   No      NaN  \n",
       "4  Mergaert, J., Ruffieux, K., Bourban, C., Storm...      No   No      NaN  \n",
       "5  Morohoshi, T., Oi, T., Aiso, H., Suzuki, T., O...      No   No      NaN  \n",
       "6  Uchida, H., Nakajima-Kambe, T., Shigeno-Akutsu...  Lipase   No      NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#If it returns an error, just wait, it will get the dataframe anyway.\n",
    "for index, row in df.iterrows():\n",
    "    if row[\"Tax ID\"] == 0:\n",
    "        name2taxid = ncbi.get_name_translator([row[\"Microorganism\"]])\n",
    "        df[\"Tax ID\"][row.name] = str(name2taxid.values()).replace(\"[\",\"\").replace(\"]\",\"\").split(\"(\")[1].split(\")\")[0]\n",
    "df.head(7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-dataframe-with-all-microorganisms-in-the-rows-and-the-plastics-in-the-columns'></a>\n",
    "# Creating dataframe with all microorganisms in the rows and the plastics in the columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BTA copolyester</th>\n",
       "      <th>Ecoflex</th>\n",
       "      <th>HDPE</th>\n",
       "      <th>LDPE</th>\n",
       "      <th>LDPE Blend</th>\n",
       "      <th>Mater-Bi</th>\n",
       "      <th>Nylon</th>\n",
       "      <th>Nylon 12</th>\n",
       "      <th>Nylon 4</th>\n",
       "      <th>Nylon 6</th>\n",
       "      <th>...</th>\n",
       "      <th>PTMS</th>\n",
       "      <th>PTS</th>\n",
       "      <th>PU</th>\n",
       "      <th>PU Blend</th>\n",
       "      <th>PVA-LLDPE</th>\n",
       "      <th>PVC</th>\n",
       "      <th>PVC Blend</th>\n",
       "      <th>Poly(3HB-co-4HB)</th>\n",
       "      <th>Polyether sulfone</th>\n",
       "      <th>Sky-Green</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     BTA copolyester  Ecoflex  HDPE  LDPE  LDPE Blend  Mater-Bi  Nylon  \\\n",
       "232                0        0     0     0           0         0      0   \n",
       "239                0        0     0     0           0         0      0   \n",
       "285                0        0     0     0           0         0      0   \n",
       "287                0        0     0     0           0         0      0   \n",
       "292                0        0     0     0           0         0      0   \n",
       "294                0        0     0     0           0         0      0   \n",
       "300                0        0     0     0           0         0      0   \n",
       "\n",
       "     Nylon 12  Nylon 4  Nylon 6  ...  PTMS  PTS  PU  PU Blend  PVA-LLDPE  PVC  \\\n",
       "232         0        0        0  ...     0    0   0         0          0    0   \n",
       "239         0        0        0  ...     0    0   0         0          0    0   \n",
       "285         0        0        0  ...     0    0   0         0          0    0   \n",
       "287         0        0        0  ...     0    0   0         0          0    0   \n",
       "292         0        0        0  ...     0    0   0         0          0    0   \n",
       "294         0        0        0  ...     0    0   0         0          0    0   \n",
       "300         0        0        0  ...     0    0   0         0          0    0   \n",
       "\n",
       "     PVC Blend  Poly(3HB-co-4HB)  Polyether sulfone  Sky-Green  \n",
       "232          0                 0                  0          0  \n",
       "239          0                 0                  0          0  \n",
       "285          0                 0                  0          0  \n",
       "287          0                 0                  0          0  \n",
       "292          0                 0                  0          0  \n",
       "294          0                 0                  0          0  \n",
       "300          0                 0                  0          0  \n",
       "\n",
       "[7 rows x 63 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plastic_types = df['Plastic'].unique()\n",
    "tax_ids = df['Tax ID'].unique()\n",
    "df2 = pd.DataFrame(0, index=tax_ids, columns=plastic_types)\n",
    "df2.sort_index(inplace=True)\n",
    "df2.sort_index(axis=1, inplace=True)\n",
    "df2.head(7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Filling-the-dataframe-with-plastic-degrading-data'></a>\n",
    "# Filling the dataframe with plastic degrading data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BTA copolyester</th>\n",
       "      <th>Ecoflex</th>\n",
       "      <th>HDPE</th>\n",
       "      <th>LDPE</th>\n",
       "      <th>LDPE Blend</th>\n",
       "      <th>Mater-Bi</th>\n",
       "      <th>Nylon</th>\n",
       "      <th>Nylon 12</th>\n",
       "      <th>Nylon 4</th>\n",
       "      <th>Nylon 6</th>\n",
       "      <th>...</th>\n",
       "      <th>PTMS</th>\n",
       "      <th>PTS</th>\n",
       "      <th>PU</th>\n",
       "      <th>PU Blend</th>\n",
       "      <th>PVA-LLDPE</th>\n",
       "      <th>PVC</th>\n",
       "      <th>PVC Blend</th>\n",
       "      <th>Poly(3HB-co-4HB)</th>\n",
       "      <th>Polyether sulfone</th>\n",
       "      <th>Sky-Green</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     BTA copolyester  Ecoflex  HDPE  LDPE  LDPE Blend  Mater-Bi  Nylon  \\\n",
       "232                0        0     0     0           0         0      0   \n",
       "239                0        0     0     0           0         0      0   \n",
       "285                0        0     0     0           0         0      0   \n",
       "287                0        0     1     1           0         0      0   \n",
       "292                0        0     0     0           0         0      0   \n",
       "294                0        0     1     0           0         0      0   \n",
       "300                0        0     0     0           0         0      0   \n",
       "\n",
       "     Nylon 12  Nylon 4  Nylon 6  ...  PTMS  PTS  PU  PU Blend  PVA-LLDPE  PVC  \\\n",
       "232         0        0        0  ...     0    0   0         0          0    0   \n",
       "239         0        0        0  ...     0    0   0         0          0    0   \n",
       "285         0        0        0  ...     0    0   0         0          0    0   \n",
       "287         0        0        0  ...     0    0   1         1          0    0   \n",
       "292         0        0        0  ...     0    0   0         0          0    0   \n",
       "294         0        0        0  ...     0    0   0         0          0    1   \n",
       "300         0        0        0  ...     0    0   0         0          0    0   \n",
       "\n",
       "     PVC Blend  Poly(3HB-co-4HB)  Polyether sulfone  Sky-Green  \n",
       "232          0                 0                  0          0  \n",
       "239          0                 0                  0          0  \n",
       "285          0                 0                  0          0  \n",
       "287          0                 0                  0          0  \n",
       "292          0                 0                  0          0  \n",
       "294          0                 0                  0          0  \n",
       "300          0                 0                  0          0  \n",
       "\n",
       "[7 rows x 63 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references_dict = {}\n",
    "for index, row in df.iterrows():\n",
    "    plastic = row['Plastic']\n",
    "    tax_id = row['Tax ID']\n",
    "    reference = str(row['Ref'].encode('utf-8').strip())\n",
    "    df2.at[tax_id, plastic] = 1\n",
    "    if tax_id not in references_dict.keys():\n",
    "        references_dict[tax_id] = reference\n",
    "    else:\n",
    "        references_dict[tax_id] += reference\n",
    "df2.head(7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Getting-tree-topology-from-NCBI-and-creating-the-tree'></a>\n",
    "# Getting tree topology from NCBI and creating the tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = df2.index.values\n",
    "t = ncbi.get_topology(ids, intermediate_nodes=True)\n",
    "'''\n",
    "leaves = t.get_leaves()\n",
    "for n in range(0, len(leaves)):\n",
    "    tax_id = int(leaves[n].name)\n",
    "    reference = references_dict[tax_id]\n",
    "    leaves[n].add_feature(\"References\", reference)\n",
    "    taxid_to_sci_name_dict[leaves[n].name] = leaves[n].sci_name\n",
    "    leaves[n].name = leaves[n].sci_name\n",
    "'''\n",
    "t.write(format=8, outfile=\"tree.nw\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-dictionary-with-organism-names-to-TAX-IDs'></a>\n",
    "# Creating dictionary with organism names to TAX IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_id_to_organism_name_dict = {}\n",
    "for index, row in df2.iterrows():\n",
    "    tax_id_to_organism_name_dict[index] = taxid_to_sci_name_dict[str(index)].replace(\"(\",\"_\").replace(\":\",\"_\").replace(\")\",\"_\").replace(\"[\",\"_\").replace(\"]\",\"_\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-Itol-dataset-files'></a>\n",
    "# Creating Itol dataset files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = lambda: random.randint(0,255) #Creates random hexadecimal color codes\n",
    "create_directory(\"itol_datasets\")\n",
    "\n",
    "for column in df2.columns:\n",
    "    f = open(\"itol_datasets/\"+column.rstrip()+\".txt\", \"w\")\n",
    "    f.write(\"\"\"DATASET_BINARY\n",
    "SEPARATOR COMMA\n",
    "DATASET_LABEL,\"\"\"+column+\"\"\"\n",
    "COLOR,\"\"\"+'#%02X%02X%02X' % (r(),r(),r())+\"\"\"\n",
    "FIELD_SHAPES,1\n",
    "FIELD_LABELS,f1\n",
    "DATA\"\"\")\n",
    "    for index, row in df2.iterrows():\n",
    "        tax_id = str(index)\n",
    "        feature_binary = str(row[column])\n",
    "        f.write(\"\\n\"+tax_id+\",\"+feature_binary)\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Downloading-all-available-genomes-for-the-plastic-degraders'></a>\n",
    "# Downloading all available genomes for the plastic degraders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "for tax_id in tax_ids:\n",
    "    job_name = str(tax_id)+\"_download\"\n",
    "    job_time = \"05:00:00\"\n",
    "    job_memory = 100 #Amount of ram memory in MB\n",
    "    job_cpus = 1\n",
    "    job_command = \"ncbi-genome-download -s genbank --format fasta --taxid \"+str(tax_id)+\" -r 5 -o genomes/\"+str(tax_id)+\"/ fungi,bacteria\"\n",
    "    run_command(run_on_slurm, account, job_name, job_time, job_memory, job_cpus, job_command)\n",
    "    time.sleep(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-itol-dataset-for-microorganisms-with-available-genomes'></a>\n",
    "# Creating itol dataset for microorganisms with available genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "create_directory(\"itol_datasets_genomes\")\n",
    "directories_list = os.listdir(\"genomes\")\n",
    "\n",
    "f = open(\"itol_datasets_genomes/Available_genomes.txt\", \"w\")\n",
    "f.write(\"\"\"DATASET_BINARY\n",
    "SEPARATOR COMMA\n",
    "DATASET_LABEL,Available genomes\n",
    "COLOR,#FF0000\n",
    "FIELD_SHAPES,1\n",
    "FIELD_LABELS,f1\n",
    "DATA\"\"\")\n",
    "\n",
    "for index, row in df2.iterrows():\n",
    "    organism_name = tax_id_to_organism_name_dict[index]\n",
    "    if str(index) in directories_list:\n",
    "        f.write(\"\\n\"+organism_name+\",\"+\"1\")\n",
    "    else:\n",
    "        f.write(\"\\n\"+organism_name+\",\"+\"0\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Blasting-genes-against-genomes'></a>\n",
    "# Blasting genes against genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "\n",
    "#Adding the TAX ID to all sequence names\n",
    "create_directory(\"genomes_database\")\n",
    "os.system(\"touch genomes_database/genomes.fasta\")\n",
    "for directory in directories_list:\n",
    "    os.system(\"zcat genomes/\"+directory+\"/genbank/*/*/*genomic.fna.gz | sed s/'^>'/'>\"+directory+\"_\"+\"'/g >> genomes_database/genomes.fasta\")\n",
    "\n",
    "#Creating BLAST database\n",
    "os.system(\"makeblastdb -in genomes_database/genomes.fasta -dbtype nucl\")\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "#Sending BLAST jobs\n",
    "create_directory(\"blasts\")\n",
    "job_time = \"01:00:00\"\n",
    "job_memory = 512 #Amount of ram memory in MB\n",
    "job_cpus = 1\n",
    "for index, row in df.iterrows():\n",
    "    sequences = row[\"Sequence\"]\n",
    "    if pd.notna(sequences) == True:\n",
    "        for n in range(0, len(sequences.split(\";\"))):\n",
    "            species_name = tax_id_to_organism_name_dict[row[\"Tax ID\"]].replace(\" \",\"_\")\n",
    "            job_name = str(row[\"Tax ID\"])+\"_\"+str(n)\n",
    "            job_command = \"echo '\"+sequences.split(\";\")[n]+\"' | tblastn -query - -db genomes_database/genomes.fasta -out blasts/\"+species_name+\"_\"+row[\"Enzyme\"].replace(\" \",\"_\")+\"_\"+str(n)+\".tsv -outfmt 6\"\n",
    "            run_command(run_on_slurm, account, job_name, job_time, job_memory, job_cpus, job_command)\n",
    "            time.sleep(2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-itol-datasets-for-all-genes'></a>\n",
    "# Creating itol datasets for all genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = lambda: random.randint(0,255) #Creates random hexadecimal color codes\n",
    "\n",
    "blast_outputs_list = os.listdir(\"blasts\")\n",
    "for file in blast_outputs_list:\n",
    "    file_blast = open(\"blasts/\"+file)\n",
    "    blast = file_blast.readlines()\n",
    "    file_itol = open(\"itol_datasets_genomes/\"+file.replace(\"tsv\", \"txt\"), \"w\")\n",
    "    counting_dict = {}\n",
    "    for line in blast:\n",
    "        qseqid = line.rstrip().split(\"\\t\")[0]\n",
    "        sseqid = line.rstrip().split(\"\\t\")[1].split(\"_\")[0]\n",
    "        pident = float(line.rstrip().split(\"\\t\")[2])\n",
    "        length = int(line.rstrip().split(\"\\t\")[3])\n",
    "        mismatch = int(line.rstrip().split(\"\\t\")[4])\n",
    "        gapopen = int(line.rstrip().split(\"\\t\")[5])\n",
    "        qstart = int(line.rstrip().split(\"\\t\")[6])\n",
    "        qend = float(line.rstrip().split(\"\\t\")[7])\n",
    "        sstart = float(line.rstrip().split(\"\\t\")[8])\n",
    "        send = float(line.rstrip().split(\"\\t\")[9])\n",
    "        evalue = float(line.rstrip().split(\"\\t\")[10])\n",
    "        bitscore = float(line.rstrip().split(\"\\t\")[11])\n",
    "        if evalue <= 1e-06 and pident >= 25:\n",
    "            if sseqid not in counting_dict:\n",
    "                counting_dict[sseqid] = 1\n",
    "            else:\n",
    "                counting_dict[sseqid] += 1\n",
    "                \n",
    "    file_itol.write(\"\"\"DATASET_MULTIBAR\n",
    "SEPARATOR TAB\n",
    "DATASET_LABEL\t\"\"\"+file.replace('.tsv', '').replace('_', ' ')+\"\"\"\n",
    "COLOR\t\"\"\"+'#%02X%02X%02X' % (r(),r(),r())+\"\"\"\n",
    "FIELD_COLORS\t\"\"\"+'#%02X%02X%02X' % (r(),r(),r())+\"\"\"\n",
    "FIELD_LABELS\t\"\"\"+file.replace('.tsv', '').replace('_', ' ')+\"\"\"\n",
    "DATA\"\"\")    \n",
    "    for taxid, organism_name in tax_id_to_organism_name_dict.items():\n",
    "        if str(taxid) in counting_dict:\n",
    "            file_itol.write(\"\\n\"+organism_name+\"\\t\"+str(counting_dict[str(taxid)]))\n",
    "        else:\n",
    "            file_itol.write(\"\\n\"+organism_name+\"\\t0\")\n",
    "    file_blast.close()\n",
    "    file_itol.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='Creating-itol-datasets-for-phylum-labels'></a>\n",
    "# Creating itol datasets for phylum labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gammaproteobacteria [232, 287, 294, 300, 303, 306, 316, 317, 470, 471, 479, 549, 573, 615, 663, 670, 29446, 34038, 35703, 40215, 40324, 42895, 43263, 47877, 47878, 47881, 47886, 48296, 50422, 53408, 56811, 61645, 61652, 69392, 69393, 76759, 76760, 76761, 78556, 108980, 137658, 202952, 226011, 254161, 319939, 380021, 405444, 477689, 570156, 587753, 676599, 693231, 1148157, 1407621, 1869214, 1871049, 1872427]\n",
      "Flavobacteriia [239]\n",
      "Betaproteobacteria [285, 292, 329, 511, 512, 12917, 28066, 28095, 29443, 34028, 34030, 34073, 36773, 47920, 54061, 60552, 65048, 75659, 76731, 80866, 80867, 85698, 94624, 106590, 180282, 215580, 401470, 592050, 1547922, 1872122, 1872437, 1873897, 1886637, 1934313]\n",
      "Alphaproteobacteria [358, 1673, 42190, 94625, 335286]\n",
      "Oligoflexia [959]\n",
      "Actinobacteria [1270, 1271, 1273, 1667, 1720, 1785, 1814, 1816, 1824, 1829, 1830, 1831, 1833, 1852, 1877, 1881, 1911, 1931, 1934, 1941, 1952, 1989, 2017, 2021, 2030, 2065, 2072, 31958, 33907, 33910, 33919, 33920, 33921, 36819, 37632, 40571, 43354, 43355, 46173, 47763, 51671, 53356, 53522, 58117, 58343, 60443, 60912, 66867, 67313, 67342, 67581, 68192, 71999, 75292, 76020, 76021, 76022, 84724, 103731, 103733, 103734, 137662, 199592, 208445, 295352, 310350, 409290, 415216, 476551, 547461, 589330, 1871184, 1871614, 1898440]\n",
      "Bacilli [1282, 1288, 1306, 1390, 1391, 1393, 1396, 1397, 1402, 1404, 1405, 1408, 1409, 1421, 1422, 1423, 1428, 1451, 1479, 1482, 2026, 28031, 29382, 29387, 33938, 33941, 35841, 37482, 44252, 44751, 45462, 51101, 54914, 54915, 58172, 64104, 72361, 86664, 169177, 293386, 293387, 324768, 332410, 412384, 521520, 859144, 1871618, 1871623, 1882945, 1960874]\n",
      "Clostridia [1488, 1491]\n",
      "Saccharomycetes [4929, 4959, 5045, 5481]\n",
      "Eurotiomycetes [5057, 5059, 5061, 5062, 5065, 5068, 5076, 5078, 5079, 5081, 5082, 5087, 28565, 28572, 28574, 28576, 33178, 34381, 36630, 40382, 40383, 40412, 41061, 41413, 41959, 46472, 63131, 63820, 63821, 69488, 69763, 69773, 69780, 69781, 69784, 69808, 75750, 105351, 128442, 162425, 198730, 264951, 454130, 746128, 1963841]\n",
      "Agaricomycetes [5306, 5322, 5325, 5331, 5341, 5365, 119078, 155003, 181124]\n",
      "Tremellomycetes [5418]\n",
      "Sordariomycetes [5507, 5516, 5530, 5544, 5547, 29856, 29875, 29916, 33203, 36460, 37998, 38033, 45245, 45277, 48865, 61235, 64636, 85828, 89141, 93594, 114497, 117187, 169388, 226117, 929575, 1105325, 1136066, 1715253, 1769349, 1981612, 2046025]\n",
      "Dothideomycetes [5580, 5599, 5791, 29917, 48100, 143450, 215132, 437017, 671092, 798071, 1707700, 1707701, 1715220, 1755431, 1755435, 1836197, 1849494, 1914427, 1970644]\n",
      "Sphingobacteriia [28454]\n",
      "Mucoromycetes [36080, 64647, 936053, 1715236, 1981611, 1982014, 2044285]\n",
      "Mortierellomycetes [64518, 979712, 1715235]\n",
      "Leotiomycetes [79858]\n",
      "Ustilaginomycetes [84753, 1902915]\n",
      "Microbotryomycetes [86839]\n",
      "Negativicutes [1871332]\n",
      "Gammaproteobacteria [232, 287, 294, 300, 303, 306, 316, 317, 470, 471, 479, 549, 573, 615, 663, 670, 29446, 34038, 35703, 40215, 40324, 42895, 43263, 47877, 47878, 47881, 47886, 48296, 50422, 53408, 56811, 61645, 61652, 69392, 69393, 76759, 76760, 76761, 78556, 108980, 137658, 202952, 226011, 254161, 319939, 380021, 405444, 477689, 570156, 587753, 676599, 693231, 1148157, 1407621, 1869214, 1871049, 1872427]\n",
      "Flavobacteriia [239]\n",
      "Betaproteobacteria [285, 292, 329, 511, 512, 12917, 28066, 28095, 29443, 34028, 34030, 34073, 36773, 47920, 54061, 60552, 65048, 75659, 76731, 80866, 80867, 85698, 94624, 106590, 180282, 215580, 401470, 592050, 1547922, 1872122, 1872437, 1873897, 1886637, 1934313]\n",
      "Alphaproteobacteria [358, 1673, 42190, 94625, 335286]\n",
      "Oligoflexia [959]\n",
      "Actinobacteria [1270, 1271, 1273, 1667, 1720, 1785, 1814, 1816, 1824, 1829, 1830, 1831, 1833, 1852, 1877, 1881, 1911, 1931, 1934, 1941, 1952, 1989, 2017, 2021, 2030, 2065, 2072, 31958, 33907, 33910, 33919, 33920, 33921, 36819, 37632, 40571, 43354, 43355, 46173, 47763, 51671, 53356, 53522, 58117, 58343, 60443, 60912, 66867, 67313, 67342, 67581, 68192, 71999, 75292, 76020, 76021, 76022, 84724, 103731, 103733, 103734, 137662, 199592, 208445, 295352, 310350, 409290, 415216, 476551, 547461, 589330, 1871184, 1871614, 1898440]\n",
      "Bacilli [1282, 1288, 1306, 1390, 1391, 1393, 1396, 1397, 1402, 1404, 1405, 1408, 1409, 1421, 1422, 1423, 1428, 1451, 1479, 1482, 2026, 28031, 29382, 29387, 33938, 33941, 35841, 37482, 44252, 44751, 45462, 51101, 54914, 54915, 58172, 64104, 72361, 86664, 169177, 293386, 293387, 324768, 332410, 412384, 521520, 859144, 1871618, 1871623, 1882945, 1960874]\n",
      "Clostridia [1488, 1491]\n",
      "Saccharomycetes [4929, 4959, 5045, 5481]\n",
      "Eurotiomycetes [5057, 5059, 5061, 5062, 5065, 5068, 5076, 5078, 5079, 5081, 5082, 5087, 28565, 28572, 28574, 28576, 33178, 34381, 36630, 40382, 40383, 40412, 41061, 41413, 41959, 46472, 63131, 63820, 63821, 69488, 69763, 69773, 69780, 69781, 69784, 69808, 75750, 105351, 128442, 162425, 198730, 264951, 454130, 746128, 1963841]\n",
      "Agaricomycetes [5306, 5322, 5325, 5331, 5341, 5365, 119078, 155003, 181124]\n",
      "Tremellomycetes [5418]\n",
      "Sordariomycetes [5507, 5516, 5530, 5544, 5547, 29856, 29875, 29916, 33203, 36460, 37998, 38033, 45245, 45277, 48865, 61235, 64636, 85828, 89141, 93594, 114497, 117187, 169388, 226117, 929575, 1105325, 1136066, 1715253, 1769349, 1981612, 2046025]\n",
      "Dothideomycetes [5580, 5599, 5791, 29917, 48100, 143450, 215132, 437017, 671092, 798071, 1707700, 1707701, 1715220, 1755431, 1755435, 1836197, 1849494, 1914427, 1970644]\n",
      "Sphingobacteriia [28454]\n",
      "Mucoromycetes [36080, 64647, 936053, 1715236, 1981611, 1982014, 2044285]\n",
      "Mortierellomycetes [64518, 979712, 1715235]\n",
      "Leotiomycetes [79858]\n",
      "Ustilaginomycetes [84753, 1902915]\n",
      "Microbotryomycetes [86839]\n",
      "Negativicutes [1871332]\n",
      "Gammaproteobacteria [232, 287, 294, 300, 303, 306, 316, 317, 470, 471, 479, 549, 573, 615, 663, 670, 29446, 34038, 35703, 40215, 40324, 42895, 43263, 47877, 47878, 47881, 47886, 48296, 50422, 53408, 56811, 61645, 61652, 69392, 69393, 76759, 76760, 76761, 78556, 108980, 137658, 202952, 226011, 254161, 319939, 380021, 405444, 477689, 570156, 587753, 676599, 693231, 1148157, 1407621, 1869214, 1871049, 1872427]\n",
      "Flavobacteriia [239]\n",
      "Betaproteobacteria [285, 292, 329, 511, 512, 12917, 28066, 28095, 29443, 34028, 34030, 34073, 36773, 47920, 54061, 60552, 65048, 75659, 76731, 80866, 80867, 85698, 94624, 106590, 180282, 215580, 401470, 592050, 1547922, 1872122, 1872437, 1873897, 1886637, 1934313]\n",
      "Alphaproteobacteria [358, 1673, 42190, 94625, 335286]\n",
      "Oligoflexia [959]\n",
      "Actinobacteria [1270, 1271, 1273, 1667, 1720, 1785, 1814, 1816, 1824, 1829, 1830, 1831, 1833, 1852, 1877, 1881, 1911, 1931, 1934, 1941, 1952, 1989, 2017, 2021, 2030, 2065, 2072, 31958, 33907, 33910, 33919, 33920, 33921, 36819, 37632, 40571, 43354, 43355, 46173, 47763, 51671, 53356, 53522, 58117, 58343, 60443, 60912, 66867, 67313, 67342, 67581, 68192, 71999, 75292, 76020, 76021, 76022, 84724, 103731, 103733, 103734, 137662, 199592, 208445, 295352, 310350, 409290, 415216, 476551, 547461, 589330, 1871184, 1871614, 1898440]\n",
      "Bacilli [1282, 1288, 1306, 1390, 1391, 1393, 1396, 1397, 1402, 1404, 1405, 1408, 1409, 1421, 1422, 1423, 1428, 1451, 1479, 1482, 2026, 28031, 29382, 29387, 33938, 33941, 35841, 37482, 44252, 44751, 45462, 51101, 54914, 54915, 58172, 64104, 72361, 86664, 169177, 293386, 293387, 324768, 332410, 412384, 521520, 859144, 1871618, 1871623, 1882945, 1960874]\n",
      "Clostridia [1488, 1491]\n",
      "Saccharomycetes [4929, 4959, 5045, 5481]\n",
      "Eurotiomycetes [5057, 5059, 5061, 5062, 5065, 5068, 5076, 5078, 5079, 5081, 5082, 5087, 28565, 28572, 28574, 28576, 33178, 34381, 36630, 40382, 40383, 40412, 41061, 41413, 41959, 46472, 63131, 63820, 63821, 69488, 69763, 69773, 69780, 69781, 69784, 69808, 75750, 105351, 128442, 162425, 198730, 264951, 454130, 746128, 1963841]\n",
      "Agaricomycetes [5306, 5322, 5325, 5331, 5341, 5365, 119078, 155003, 181124]\n",
      "Tremellomycetes [5418]\n",
      "Sordariomycetes [5507, 5516, 5530, 5544, 5547, 29856, 29875, 29916, 33203, 36460, 37998, 38033, 45245, 45277, 48865, 61235, 64636, 85828, 89141, 93594, 114497, 117187, 169388, 226117, 929575, 1105325, 1136066, 1715253, 1769349, 1981612, 2046025]\n",
      "Dothideomycetes [5580, 5599, 5791, 29917, 48100, 143450, 215132, 437017, 671092, 798071, 1707700, 1707701, 1715220, 1755431, 1755435, 1836197, 1849494, 1914427, 1970644]\n",
      "Sphingobacteriia [28454]\n",
      "Mucoromycetes [36080, 64647, 936053, 1715236, 1981611, 1982014, 2044285]\n",
      "Mortierellomycetes [64518, 979712, 1715235]\n",
      "Leotiomycetes [79858]\n",
      "Ustilaginomycetes [84753, 1902915]\n",
      "Microbotryomycetes [86839]\n",
      "Negativicutes [1871332]\n",
      "Gammaproteobacteria [232, 287, 294, 300, 303, 306, 316, 317, 470, 471, 479, 549, 573, 615, 663, 670, 29446, 34038, 35703, 40215, 40324, 42895, 43263, 47877, 47878, 47881, 47886, 48296, 50422, 53408, 56811, 61645, 61652, 69392, 69393, 76759, 76760, 76761, 78556, 108980, 137658, 202952, 226011, 254161, 319939, 380021, 405444, 477689, 570156, 587753, 676599, 693231, 1148157, 1407621, 1869214, 1871049, 1872427]\n",
      "Flavobacteriia [239]\n",
      "Betaproteobacteria [285, 292, 329, 511, 512, 12917, 28066, 28095, 29443, 34028, 34030, 34073, 36773, 47920, 54061, 60552, 65048, 75659, 76731, 80866, 80867, 85698, 94624, 106590, 180282, 215580, 401470, 592050, 1547922, 1872122, 1872437, 1873897, 1886637, 1934313]\n",
      "Alphaproteobacteria [358, 1673, 42190, 94625, 335286]\n",
      "Oligoflexia [959]\n",
      "Actinobacteria [1270, 1271, 1273, 1667, 1720, 1785, 1814, 1816, 1824, 1829, 1830, 1831, 1833, 1852, 1877, 1881, 1911, 1931, 1934, 1941, 1952, 1989, 2017, 2021, 2030, 2065, 2072, 31958, 33907, 33910, 33919, 33920, 33921, 36819, 37632, 40571, 43354, 43355, 46173, 47763, 51671, 53356, 53522, 58117, 58343, 60443, 60912, 66867, 67313, 67342, 67581, 68192, 71999, 75292, 76020, 76021, 76022, 84724, 103731, 103733, 103734, 137662, 199592, 208445, 295352, 310350, 409290, 415216, 476551, 547461, 589330, 1871184, 1871614, 1898440]\n",
      "Bacilli [1282, 1288, 1306, 1390, 1391, 1393, 1396, 1397, 1402, 1404, 1405, 1408, 1409, 1421, 1422, 1423, 1428, 1451, 1479, 1482, 2026, 28031, 29382, 29387, 33938, 33941, 35841, 37482, 44252, 44751, 45462, 51101, 54914, 54915, 58172, 64104, 72361, 86664, 169177, 293386, 293387, 324768, 332410, 412384, 521520, 859144, 1871618, 1871623, 1882945, 1960874]\n",
      "Clostridia [1488, 1491]\n",
      "Saccharomycetes [4929, 4959, 5045, 5481]\n",
      "Eurotiomycetes [5057, 5059, 5061, 5062, 5065, 5068, 5076, 5078, 5079, 5081, 5082, 5087, 28565, 28572, 28574, 28576, 33178, 34381, 36630, 40382, 40383, 40412, 41061, 41413, 41959, 46472, 63131, 63820, 63821, 69488, 69763, 69773, 69780, 69781, 69784, 69808, 75750, 105351, 128442, 162425, 198730, 264951, 454130, 746128, 1963841]\n",
      "Agaricomycetes [5306, 5322, 5325, 5331, 5341, 5365, 119078, 155003, 181124]\n",
      "Tremellomycetes [5418]\n",
      "Sordariomycetes [5507, 5516, 5530, 5544, 5547, 29856, 29875, 29916, 33203, 36460, 37998, 38033, 45245, 45277, 48865, 61235, 64636, 85828, 89141, 93594, 114497, 117187, 169388, 226117, 929575, 1105325, 1136066, 1715253, 1769349, 1981612, 2046025]\n",
      "Dothideomycetes [5580, 5599, 5791, 29917, 48100, 143450, 215132, 437017, 671092, 798071, 1707700, 1707701, 1715220, 1755431, 1755435, 1836197, 1849494, 1914427, 1970644]\n",
      "Sphingobacteriia [28454]\n",
      "Mucoromycetes [36080, 64647, 936053, 1715236, 1981611, 1982014, 2044285]\n",
      "Mortierellomycetes [64518, 979712, 1715235]\n",
      "Leotiomycetes [79858]\n",
      "Ustilaginomycetes [84753, 1902915]\n",
      "Microbotryomycetes [86839]\n",
      "Negativicutes [1871332]\n",
      "Gammaproteobacteria [232, 287, 294, 300, 303, 306, 316, 317, 470, 471, 479, 549, 573, 615, 663, 670, 29446, 34038, 35703, 40215, 40324, 42895, 43263, 47877, 47878, 47881, 47886, 48296, 50422, 53408, 56811, 61645, 61652, 69392, 69393, 76759, 76760, 76761, 78556, 108980, 137658, 202952, 226011, 254161, 319939, 380021, 405444, 477689, 570156, 587753, 676599, 693231, 1148157, 1407621, 1869214, 1871049, 1872427]\n",
      "Flavobacteriia [239]\n",
      "Betaproteobacteria [285, 292, 329, 511, 512, 12917, 28066, 28095, 29443, 34028, 34030, 34073, 36773, 47920, 54061, 60552, 65048, 75659, 76731, 80866, 80867, 85698, 94624, 106590, 180282, 215580, 401470, 592050, 1547922, 1872122, 1872437, 1873897, 1886637, 1934313]\n",
      "Alphaproteobacteria [358, 1673, 42190, 94625, 335286]\n",
      "Oligoflexia [959]\n",
      "Actinobacteria [1270, 1271, 1273, 1667, 1720, 1785, 1814, 1816, 1824, 1829, 1830, 1831, 1833, 1852, 1877, 1881, 1911, 1931, 1934, 1941, 1952, 1989, 2017, 2021, 2030, 2065, 2072, 31958, 33907, 33910, 33919, 33920, 33921, 36819, 37632, 40571, 43354, 43355, 46173, 47763, 51671, 53356, 53522, 58117, 58343, 60443, 60912, 66867, 67313, 67342, 67581, 68192, 71999, 75292, 76020, 76021, 76022, 84724, 103731, 103733, 103734, 137662, 199592, 208445, 295352, 310350, 409290, 415216, 476551, 547461, 589330, 1871184, 1871614, 1898440]\n",
      "Bacilli [1282, 1288, 1306, 1390, 1391, 1393, 1396, 1397, 1402, 1404, 1405, 1408, 1409, 1421, 1422, 1423, 1428, 1451, 1479, 1482, 2026, 28031, 29382, 29387, 33938, 33941, 35841, 37482, 44252, 44751, 45462, 51101, 54914, 54915, 58172, 64104, 72361, 86664, 169177, 293386, 293387, 324768, 332410, 412384, 521520, 859144, 1871618, 1871623, 1882945, 1960874]\n",
      "Clostridia [1488, 1491]\n",
      "Saccharomycetes [4929, 4959, 5045, 5481]\n",
      "Eurotiomycetes [5057, 5059, 5061, 5062, 5065, 5068, 5076, 5078, 5079, 5081, 5082, 5087, 28565, 28572, 28574, 28576, 33178, 34381, 36630, 40382, 40383, 40412, 41061, 41413, 41959, 46472, 63131, 63820, 63821, 69488, 69763, 69773, 69780, 69781, 69784, 69808, 75750, 105351, 128442, 162425, 198730, 264951, 454130, 746128, 1963841]\n",
      "Agaricomycetes [5306, 5322, 5325, 5331, 5341, 5365, 119078, 155003, 181124]\n",
      "Tremellomycetes [5418]\n",
      "Sordariomycetes [5507, 5516, 5530, 5544, 5547, 29856, 29875, 29916, 33203, 36460, 37998, 38033, 45245, 45277, 48865, 61235, 64636, 85828, 89141, 93594, 114497, 117187, 169388, 226117, 929575, 1105325, 1136066, 1715253, 1769349, 1981612, 2046025]\n",
      "Dothideomycetes [5580, 5599, 5791, 29917, 48100, 143450, 215132, 437017, 671092, 798071, 1707700, 1707701, 1715220, 1755431, 1755435, 1836197, 1849494, 1914427, 1970644]\n",
      "Sphingobacteriia [28454]\n",
      "Mucoromycetes [36080, 64647, 936053, 1715236, 1981611, 1982014, 2044285]\n",
      "Mortierellomycetes [64518, 979712, 1715235]\n",
      "Leotiomycetes [79858]\n",
      "Ustilaginomycetes [84753, 1902915]\n",
      "Microbotryomycetes [86839]\n",
      "Negativicutes [1871332]\n",
      "Gammaproteobacteria [232, 287, 294, 300, 303, 306, 316, 317, 470, 471, 479, 549, 573, 615, 663, 670, 29446, 34038, 35703, 40215, 40324, 42895, 43263, 47877, 47878, 47881, 47886, 48296, 50422, 53408, 56811, 61645, 61652, 69392, 69393, 76759, 76760, 76761, 78556, 108980, 137658, 202952, 226011, 254161, 319939, 380021, 405444, 477689, 570156, 587753, 676599, 693231, 1148157, 1407621, 1869214, 1871049, 1872427]\n",
      "Flavobacteriia [239]\n",
      "Betaproteobacteria [285, 292, 329, 511, 512, 12917, 28066, 28095, 29443, 34028, 34030, 34073, 36773, 47920, 54061, 60552, 65048, 75659, 76731, 80866, 80867, 85698, 94624, 106590, 180282, 215580, 401470, 592050, 1547922, 1872122, 1872437, 1873897, 1886637, 1934313]\n",
      "Alphaproteobacteria [358, 1673, 42190, 94625, 335286]\n",
      "Oligoflexia [959]\n",
      "Actinobacteria [1270, 1271, 1273, 1667, 1720, 1785, 1814, 1816, 1824, 1829, 1830, 1831, 1833, 1852, 1877, 1881, 1911, 1931, 1934, 1941, 1952, 1989, 2017, 2021, 2030, 2065, 2072, 31958, 33907, 33910, 33919, 33920, 33921, 36819, 37632, 40571, 43354, 43355, 46173, 47763, 51671, 53356, 53522, 58117, 58343, 60443, 60912, 66867, 67313, 67342, 67581, 68192, 71999, 75292, 76020, 76021, 76022, 84724, 103731, 103733, 103734, 137662, 199592, 208445, 295352, 310350, 409290, 415216, 476551, 547461, 589330, 1871184, 1871614, 1898440]\n",
      "Bacilli [1282, 1288, 1306, 1390, 1391, 1393, 1396, 1397, 1402, 1404, 1405, 1408, 1409, 1421, 1422, 1423, 1428, 1451, 1479, 1482, 2026, 28031, 29382, 29387, 33938, 33941, 35841, 37482, 44252, 44751, 45462, 51101, 54914, 54915, 58172, 64104, 72361, 86664, 169177, 293386, 293387, 324768, 332410, 412384, 521520, 859144, 1871618, 1871623, 1882945, 1960874]\n",
      "Clostridia [1488, 1491]\n",
      "Saccharomycetes [4929, 4959, 5045, 5481]\n",
      "Eurotiomycetes [5057, 5059, 5061, 5062, 5065, 5068, 5076, 5078, 5079, 5081, 5082, 5087, 28565, 28572, 28574, 28576, 33178, 34381, 36630, 40382, 40383, 40412, 41061, 41413, 41959, 46472, 63131, 63820, 63821, 69488, 69763, 69773, 69780, 69781, 69784, 69808, 75750, 105351, 128442, 162425, 198730, 264951, 454130, 746128, 1963841]\n",
      "Agaricomycetes [5306, 5322, 5325, 5331, 5341, 5365, 119078, 155003, 181124]\n",
      "Tremellomycetes [5418]\n",
      "Sordariomycetes [5507, 5516, 5530, 5544, 5547, 29856, 29875, 29916, 33203, 36460, 37998, 38033, 45245, 45277, 48865, 61235, 64636, 85828, 89141, 93594, 114497, 117187, 169388, 226117, 929575, 1105325, 1136066, 1715253, 1769349, 1981612, 2046025]\n",
      "Dothideomycetes [5580, 5599, 5791, 29917, 48100, 143450, 215132, 437017, 671092, 798071, 1707700, 1707701, 1715220, 1755431, 1755435, 1836197, 1849494, 1914427, 1970644]\n",
      "Sphingobacteriia [28454]\n",
      "Mucoromycetes [36080, 64647, 936053, 1715236, 1981611, 1982014, 2044285]\n",
      "Mortierellomycetes [64518, 979712, 1715235]\n",
      "Leotiomycetes [79858]\n",
      "Ustilaginomycetes [84753, 1902915]\n",
      "Microbotryomycetes [86839]\n",
      "Negativicutes [1871332]\n",
      "Gammaproteobacteria [232, 287, 294, 300, 303, 306, 316, 317, 470, 471, 479, 549, 573, 615, 663, 670, 29446, 34038, 35703, 40215, 40324, 42895, 43263, 47877, 47878, 47881, 47886, 48296, 50422, 53408, 56811, 61645, 61652, 69392, 69393, 76759, 76760, 76761, 78556, 108980, 137658, 202952, 226011, 254161, 319939, 380021, 405444, 477689, 570156, 587753, 676599, 693231, 1148157, 1407621, 1869214, 1871049, 1872427]\n",
      "Flavobacteriia [239]\n",
      "Betaproteobacteria [285, 292, 329, 511, 512, 12917, 28066, 28095, 29443, 34028, 34030, 34073, 36773, 47920, 54061, 60552, 65048, 75659, 76731, 80866, 80867, 85698, 94624, 106590, 180282, 215580, 401470, 592050, 1547922, 1872122, 1872437, 1873897, 1886637, 1934313]\n",
      "Alphaproteobacteria [358, 1673, 42190, 94625, 335286]\n",
      "Oligoflexia [959]\n",
      "Actinobacteria [1270, 1271, 1273, 1667, 1720, 1785, 1814, 1816, 1824, 1829, 1830, 1831, 1833, 1852, 1877, 1881, 1911, 1931, 1934, 1941, 1952, 1989, 2017, 2021, 2030, 2065, 2072, 31958, 33907, 33910, 33919, 33920, 33921, 36819, 37632, 40571, 43354, 43355, 46173, 47763, 51671, 53356, 53522, 58117, 58343, 60443, 60912, 66867, 67313, 67342, 67581, 68192, 71999, 75292, 76020, 76021, 76022, 84724, 103731, 103733, 103734, 137662, 199592, 208445, 295352, 310350, 409290, 415216, 476551, 547461, 589330, 1871184, 1871614, 1898440]\n",
      "Bacilli [1282, 1288, 1306, 1390, 1391, 1393, 1396, 1397, 1402, 1404, 1405, 1408, 1409, 1421, 1422, 1423, 1428, 1451, 1479, 1482, 2026, 28031, 29382, 29387, 33938, 33941, 35841, 37482, 44252, 44751, 45462, 51101, 54914, 54915, 58172, 64104, 72361, 86664, 169177, 293386, 293387, 324768, 332410, 412384, 521520, 859144, 1871618, 1871623, 1882945, 1960874]\n",
      "Clostridia [1488, 1491]\n",
      "Saccharomycetes [4929, 4959, 5045, 5481]\n",
      "Eurotiomycetes [5057, 5059, 5061, 5062, 5065, 5068, 5076, 5078, 5079, 5081, 5082, 5087, 28565, 28572, 28574, 28576, 33178, 34381, 36630, 40382, 40383, 40412, 41061, 41413, 41959, 46472, 63131, 63820, 63821, 69488, 69763, 69773, 69780, 69781, 69784, 69808, 75750, 105351, 128442, 162425, 198730, 264951, 454130, 746128, 1963841]\n",
      "Agaricomycetes [5306, 5322, 5325, 5331, 5341, 5365, 119078, 155003, 181124]\n",
      "Tremellomycetes [5418]\n",
      "Sordariomycetes [5507, 5516, 5530, 5544, 5547, 29856, 29875, 29916, 33203, 36460, 37998, 38033, 45245, 45277, 48865, 61235, 64636, 85828, 89141, 93594, 114497, 117187, 169388, 226117, 929575, 1105325, 1136066, 1715253, 1769349, 1981612, 2046025]\n",
      "Dothideomycetes [5580, 5599, 5791, 29917, 48100, 143450, 215132, 437017, 671092, 798071, 1707700, 1707701, 1715220, 1755431, 1755435, 1836197, 1849494, 1914427, 1970644]\n",
      "Sphingobacteriia [28454]\n",
      "Mucoromycetes [36080, 64647, 936053, 1715236, 1981611, 1982014, 2044285]\n",
      "Mortierellomycetes [64518, 979712, 1715235]\n",
      "Leotiomycetes [79858]\n",
      "Ustilaginomycetes [84753, 1902915]\n",
      "Microbotryomycetes [86839]\n",
      "Negativicutes [1871332]\n",
      "Gammaproteobacteria [232, 287, 294, 300, 303, 306, 316, 317, 470, 471, 479, 549, 573, 615, 663, 670, 29446, 34038, 35703, 40215, 40324, 42895, 43263, 47877, 47878, 47881, 47886, 48296, 50422, 53408, 56811, 61645, 61652, 69392, 69393, 76759, 76760, 76761, 78556, 108980, 137658, 202952, 226011, 254161, 319939, 380021, 405444, 477689, 570156, 587753, 676599, 693231, 1148157, 1407621, 1869214, 1871049, 1872427]\n",
      "Flavobacteriia [239]\n",
      "Betaproteobacteria [285, 292, 329, 511, 512, 12917, 28066, 28095, 29443, 34028, 34030, 34073, 36773, 47920, 54061, 60552, 65048, 75659, 76731, 80866, 80867, 85698, 94624, 106590, 180282, 215580, 401470, 592050, 1547922, 1872122, 1872437, 1873897, 1886637, 1934313]\n",
      "Alphaproteobacteria [358, 1673, 42190, 94625, 335286]\n",
      "Oligoflexia [959]\n",
      "Actinobacteria [1270, 1271, 1273, 1667, 1720, 1785, 1814, 1816, 1824, 1829, 1830, 1831, 1833, 1852, 1877, 1881, 1911, 1931, 1934, 1941, 1952, 1989, 2017, 2021, 2030, 2065, 2072, 31958, 33907, 33910, 33919, 33920, 33921, 36819, 37632, 40571, 43354, 43355, 46173, 47763, 51671, 53356, 53522, 58117, 58343, 60443, 60912, 66867, 67313, 67342, 67581, 68192, 71999, 75292, 76020, 76021, 76022, 84724, 103731, 103733, 103734, 137662, 199592, 208445, 295352, 310350, 409290, 415216, 476551, 547461, 589330, 1871184, 1871614, 1898440]\n",
      "Bacilli [1282, 1288, 1306, 1390, 1391, 1393, 1396, 1397, 1402, 1404, 1405, 1408, 1409, 1421, 1422, 1423, 1428, 1451, 1479, 1482, 2026, 28031, 29382, 29387, 33938, 33941, 35841, 37482, 44252, 44751, 45462, 51101, 54914, 54915, 58172, 64104, 72361, 86664, 169177, 293386, 293387, 324768, 332410, 412384, 521520, 859144, 1871618, 1871623, 1882945, 1960874]\n",
      "Clostridia [1488, 1491]\n",
      "Saccharomycetes [4929, 4959, 5045, 5481]\n",
      "Eurotiomycetes [5057, 5059, 5061, 5062, 5065, 5068, 5076, 5078, 5079, 5081, 5082, 5087, 28565, 28572, 28574, 28576, 33178, 34381, 36630, 40382, 40383, 40412, 41061, 41413, 41959, 46472, 63131, 63820, 63821, 69488, 69763, 69773, 69780, 69781, 69784, 69808, 75750, 105351, 128442, 162425, 198730, 264951, 454130, 746128, 1963841]\n",
      "Agaricomycetes [5306, 5322, 5325, 5331, 5341, 5365, 119078, 155003, 181124]\n",
      "Tremellomycetes [5418]\n",
      "Sordariomycetes [5507, 5516, 5530, 5544, 5547, 29856, 29875, 29916, 33203, 36460, 37998, 38033, 45245, 45277, 48865, 61235, 64636, 85828, 89141, 93594, 114497, 117187, 169388, 226117, 929575, 1105325, 1136066, 1715253, 1769349, 1981612, 2046025]\n",
      "Dothideomycetes [5580, 5599, 5791, 29917, 48100, 143450, 215132, 437017, 671092, 798071, 1707700, 1707701, 1715220, 1755431, 1755435, 1836197, 1849494, 1914427, 1970644]\n",
      "Sphingobacteriia [28454]\n",
      "Mucoromycetes [36080, 64647, 936053, 1715236, 1981611, 1982014, 2044285]\n",
      "Mortierellomycetes [64518, 979712, 1715235]\n",
      "Leotiomycetes [79858]\n",
      "Ustilaginomycetes [84753, 1902915]\n",
      "Microbotryomycetes [86839]\n",
      "Negativicutes [1871332]\n"
     ]
    }
   ],
   "source": [
    "f = open(\"itol_datasets/Phylum_labels.txt\", \"w\")\n",
    "f.write(\"\"\"DATASET_TEXT\n",
    "SEPARATOR COMMA\n",
    "DATASET_LABEL,example text dataset\n",
    "COLOR,#ff0000\n",
    "MARGIN,0\n",
    "SHOW_INTERNAL,1\n",
    "LABEL_ROTATION,0\n",
    "STRAIGHT_LABELS,0\n",
    "ALIGN_TO_TREE,1\n",
    "SIZE_FACTOR,1\n",
    "DATA\n",
    "\"\"\")\n",
    "\n",
    "organism_phylum_dict = {}\n",
    "organism_class_dict = {}\n",
    "\n",
    "for index, row in df2.iterrows():\n",
    "    lineage = ncbi.get_lineage(index)\n",
    "    for item in lineage:\n",
    "        if \"'phylum'\" in str(ncbi.get_rank([item])):\n",
    "            organism_phylum = ncbi.get_taxid_translator([item])[item]\n",
    "        if \"'class'\" in str(ncbi.get_rank([item])):\n",
    "            organism_class = ncbi.get_taxid_translator([item])[item]\n",
    "    if organism_class not in organism_class_dict:\n",
    "        organism_class_dict[organism_class] = [item]\n",
    "        organism_phylum_dict[organism_phylum] = [organism_class]\n",
    "    else:\n",
    "        organism_class_dict[organism_class].append(item)\n",
    "        if organism_phylum in organism_phylum_dict:\n",
    "            organism_phylum_dict[organism_phylum].append(organism_class)\n",
    "        else:\n",
    "            organism_phylum_dict[organism_phylum] = [organism_class]\n",
    "for key, value in organism_phylum_dict.items():\n",
    "    for key2, value2 in organism_class_dict.items():\n",
    "        print(key2, value2)\n",
    "        if len(value2) > 1:\n",
    "            f.write(\"\\n\"+str(value2[0])+\"|\"+str(value2[1])+\",\"+str(key)+\",-1,\"+'#%02X%02X%02X' % (r(),r(),r())+\",bold,10,0\")\n",
    "        else:\n",
    "            f.write(\"\\n\"+str(value2[0])+\"|\"+str(value2[0])+\",\"+str(key)+\",-1,\"+'#%02X%02X%02X' % (r(),r(),r())+\",bold,10,0\")\n",
    "f.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
